tools/crm_shadow.c

```diff
@@ -93,453 +145,568 @@ shadow_teardown(char *name)
+
+    crm_log_preinit(NULL, argc, argv);
+
+    if (!g_option_context_parse_strv(context, &processed_args, &error)) {
+        exit_code = CRM_EX_USAGE;
+        goto done;
     }
 
-    while (1) {
-        flag = pcmk__next_cli_option(argc, argv, &option_index, NULL);
-        if (flag == -1 || flag == 0)
-            break;
+    if (args->version) {
+        g_strfreev(processed_args);
+        pcmk__free_arg_context(context);
 
-        switch (flag) {
-            case 'a':
-                full_upload = TRUE;
-                break;
-            case 'd':
-            case 'E':
-            case 'p':
-            case 'w':
-            case 'F':
-                command = flag;
-                pcmk__str_update(&shadow, getenv("CIB_shadow"));
-                if (shadow == NULL) {
-                    fprintf(stderr, "No active shadow configuration defined\n");
-                    exit_code = CRM_EX_NOSUCH;
-                    goto done;
-                }
-                break;
-            case 'v':
-                validation = optarg;
-                break;
-            case 'e':
-            case 'c':
-            case 's':
-            case 'r':
-                command = flag;
-                pcmk__str_update(&shadow, optarg);
-                break;
-            case 'C':
-            case 'D':
-                command = flag;
-                dangerous_cmd = TRUE;
-                pcmk__str_update(&shadow, optarg);
-                break;
-            case 'V':
-                command_options = command_options | cib_verbose;
-                crm_bump_log_level(argc, argv);
-                break;
-            case '$':
-            case '?':
-                pcmk__cli_help(flag, CRM_EX_OK);
-                goto done;
-            case 'f':
-                cib__set_call_options(command_options, crm_system_name,
-                                      cib_quorum_override);
-                force_flag = 1;
-                break;
-            case 'b':
-                batch_flag = 1;
-                break;
-            default:
-                printf("Argument code 0%o (%c)" " is not (?yet?) supported\n", flag, flag);
-                ++argerr;
-                break;
-        }
+        /* FIXME: When crm_shadow is converted to use formatted output,
+         * this can go.
+         */
+        pcmk__cli_help('v', CRM_EX_OK);
     }
 
-    if (optind < argc) {
-        printf("non-option ARGV-elements: ");
-        while (optind < argc)
-            printf("%s ", argv[optind++]);
-        printf("\n");
-        pcmk__cli_help('?', CRM_EX_USAGE);
+    if (options.cmd == shadow_cmd_none) {
+        gchar *help = g_option_context_get_help(context, TRUE, NULL);
+
+        exit_code = CRM_EX_USAGE;
+        g_set_error(&error, PCMK__EXITC_ERROR, exit_code,
+                    "Must specify a query or command option\n\n%s", help);
+        g_free(help);
+        goto done;
```

Comment:
We can keep this behavior, but eventually we should be consistent across all tools whether there is a default option (some do have one)

Reply:
I agree. For now I was just keeping the `if (argc < 2)` behavior and borrowing the error message from `crmadmin`.